// mp-facil/src/errors.js
export class MercadoPagoSimplificadoError extends Error {
  constructor(code, message, options = {}) {
    super(message);
    this.name = 'MercadoPagoSimplificadoError';
    this.code = code;
    this.status = options.status;
    this.details = options.details;
    this.cause = options.cause;
  }

  toResult() {
    const result = {
      ok: false,
      erro: this.code,
      mensagem: this.message
    };

    if (this.status) result.status = this.status;
    if (this.details) result.detalhes = this.details;

    return result;
  }
}

const errorMessages = {
  API_KEY_AUSENTE: 'Informe apiKey ou api_key',
  API_KEY_INVALIDA: 'A api_key informada e invalida',
  APP_ID_INVALIDO: 'O appId informado e invalido',
  PAYLOAD_INCOMPLETO: 'Payload incompleto',
  PRODUTO_AUSENTE: 'Informe o nome do produto',
  PRECO_AUSENTE: 'Informe o preco do produto',
  PRECO_INVALIDO: 'O preco precisa ser maior que zero',
  ID_AUSENTE: 'Informe o identificador',
  FALHA_DE_REDE: 'Falha de rede ao falar com Mercado Pago',
  RESPOSTA_INESPERADA: 'Resposta inesperada da API do Mercado Pago',
  WEBHOOK_INVALIDO: 'Webhook invalido',
  ASSINATURA_INVALIDA: 'Assinatura do webhook invalida',
  STATUS_DESCONHECIDO: 'Status de pagamento nao reconhecido',
  LOOP_INTERVALO_INVALIDO: 'O intervalo do loop precisa ser maior que zero',
  LOOP_DURACAO_INVALIDA: 'A duracao do loop precisa ser maior que zero',
  LOOP_TEMPO_ESGOTADO: 'Tempo do loop esgotado antes do pagamento resolver',
  LOOP_NAO_ENCONTRADO: 'Loop de pagamento nao encontrado',
  FREQUENCIA_ASSINATURA_INVALIDA: 'A frequencia da assinatura precisa ser maior que zero',
  SPLIT_CONTA_A_INVALIDA: 'Informe o access token da conta A',
  SPLIT_CONTA_B_INVALIDA: 'Informe a porcentagem da conta B',
  SPLIT_PERCENTUAL_INVALIDO: 'As porcentagens do split precisam somar 100',
  SPLIT_VALOR_INVALIDO: 'O valor calculado do split e invalido',
  EMAIL_PAGADOR_AUSENTE: 'Informe o email do pagador para gerar Pix',
  EMAIL_PAGADOR_INVALIDO: 'O email do pagador e invalido',
  URL_RETORNO_AUSENTE: 'Informe a URL de retorno da assinatura'
};

export function createError(code, options = {}) {
  return new MercadoPagoSimplificadoError(
    code,
    options.message || errorMessages[code] || 'Erro desconhecido',
    options
  );
}

export function ok(dados) {
  return { ok: true, dados };
}

export function fail(error) {
  if (error instanceof MercadoPagoSimplificadoError) {
    return error.toResult();
  }

  return createError('RESPOSTA_INESPERADA', { cause: error }).toResult();
}
