// mp-facil/src/http-client.js
import { createError } from './errors.js';

export class HttpClient {
  constructor({ apiKey, baseUrl = 'https://api.mercadopago.com', fetchImpl = fetch }) {
    this.apiKey = apiKey;
    this.baseUrl = baseUrl;
    this.fetch = fetchImpl;
  }

  async post(path, body, options = {}) {
    return this.request('POST', path, body, options);
  }

  async get(path) {
    return this.request('GET', path);
  }

  async request(method, path, body, options = {}) {
    const apiKey = options.apiKey || this.apiKey;
    const headers = {
      Authorization: `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
      Accept: 'application/json'
    };

    if (options.idempotencyKey) {
      headers['X-Idempotency-Key'] = options.idempotencyKey;
    }

    let response;

    try {
      response = await this.fetch(`${this.baseUrl}${path}`, {
        method,
        headers,
        body: body ? JSON.stringify(body) : undefined
      });
    } catch (error) {
      throw createError('FALHA_DE_REDE', { cause: error });
    }

    const data = await this.parseResponse(response);

    if (!response.ok) {
      throw this.toApiError(response.status, data);
    }

    return data;
  }

  async parseResponse(response) {
    const text = await response.text();

    if (!text) return null;

    try {
      return JSON.parse(text);
    } catch (error) {
      throw createError('RESPOSTA_INESPERADA', {
        status: response.status,
        cause: error,
        details: text
      });
    }
  }

  toApiError(status, data) {
    const code = status === 401 ? 'API_KEY_INVALIDA' : 'RESPOSTA_INESPERADA';
    const message = data && (data.message || data.error || data.cause?.[0]?.description);

    return createError(code, {
      status,
      message: message || undefined,
      details: data
    });
  }
}
