// mp-facil/src/index.js
import { EventEmitter } from 'node:events';
import crypto from 'node:crypto';
import fs from 'node:fs/promises';
import path from 'node:path';
import { HttpClient } from './http-client.js';
import { createError, fail, ok } from './errors.js';
import {
  normalizeCharge,
  normalizeConfig,
  normalizePayment,
  normalizePixPayment,
  normalizePreference,
  normalizeRecurringPix,
  normalizeRecurringPixSubscription
} from './normalizers.js';
import { parseWebhook, paymentEventName, verifySignature } from './webhook.js';

export class MercadoPagoSimplificado extends EventEmitter {
  constructor(config = {}) {
    super();
    this.config = normalizeConfig(config);
    this.validateConfig();
    this.http = new HttpClient({
      apiKey: this.config.apiKey,
      baseUrl: this.config.baseUrl,
      fetchImpl: this.config.fetchImpl
    });
    this.memoryLoopPath = this.config.memoryLoopPath || path.join(process.cwd(), 'memory loop');
    this.loops = new Map();

    if (this.config.autoResumeLoops !== false) {
      const timer = setTimeout(() => {
        this.retomarLoops();
      }, 0);
      timer.unref?.();
    }
  }

  async criarCobranca(input) {
    return this.createCharge(input);
  }

  modoTeste(ativo = true) {
    this.config.sandbox = Boolean(ativo);
    return this;
  }

  testMode(active = true) {
    return this.modoTeste(active);
  }

  async createCharge(input) {
    try {
      const charge = normalizeCharge(input);
      this.validateCharge(charge);
      const response = await this.http.post('/checkout/preferences', this.toPreferencePayload(charge));

      return ok(normalizePreference(response, this.config.sandbox));
    } catch (error) {
      return fail(error);
    }
  }

  async criarCobrancaDividida(input) {
    return this.createSplitCharge(input);
  }

  async createSplitCharge(input) {
    try {
      const charge = normalizeCharge(input);
      this.validateCharge(charge);
      const split = this.normalizeSplit(input, charge);
      const payload = this.toPreferencePayload(charge);

      payload.marketplace_fee = split.marketplaceFee;
      payload.metadata = {
        ...(payload.metadata || {}),
        mp_facil_split: {
          contaA: split.contaA.percentual,
          contaB: split.contaB.percentual,
          marketplaceFee: split.marketplaceFee
        }
      };

      if (split.marketplace) payload.marketplace = split.marketplace;

      const response = await this.http.post('/checkout/preferences', payload, {
        apiKey: split.contaA.accessToken
      });

      return ok({
        ...normalizePreference(response, this.config.sandbox),
        split: {
          contaA: {
            percentual: split.contaA.percentual,
            valorEstimado: split.contaA.valorEstimado
          },
          contaB: {
            percentual: split.contaB.percentual,
            valor: split.marketplaceFee
          },
          observacao: 'A comissao do Mercado Pago e descontada antes da comissao do marketplace.'
        }
      });
    } catch (error) {
      return fail(error);
    }
  }

  async criarPix(input) {
    return this.createPix(input);
  }

  async createPix(input) {
    try {
      const charge = normalizeCharge(input);
      this.validateCharge(charge);
      this.validatePayerEmail(charge.payerEmail);

      const response = await this.http.post('/v1/payments', this.toPixPayload(charge), {
        idempotencyKey: this.idempotencyKey(input, charge)
      });

      return ok(normalizePixPayment(response, this.config.sandbox));
    } catch (error) {
      return fail(error);
    }
  }

  async criarPixRecorrente(input) {
    return this.createRecurringPix(input);
  }

  async createRecurringPix(input) {
    try {
      const subscription = normalizeRecurringPix(input);
      this.validateRecurringPix(subscription);

      const response = await this.http.post('/preapproval', this.toRecurringPixPayload(subscription));

      return ok(normalizeRecurringPixSubscription(response, this.config.sandbox));
    } catch (error) {
      return fail(error);
    }
  }

  async consultarCobranca(id) {
    return this.checkCharge(id);
  }

  async consultarAssinatura(id) {
    return this.checkSubscription(id);
  }

  async checkSubscription(id) {
    try {
      this.validateId(id);
      const response = await this.http.get(`/preapproval/${encodeURIComponent(id)}`);

      return ok(normalizeRecurringPixSubscription(response, this.config.sandbox));
    } catch (error) {
      return fail(error);
    }
  }

  async acompanharCobranca(id, options = {}) {
    return this.watchCharge(id, options);
  }

  async watchCharge(id, options = {}) {
    try {
      this.validateId(id);
      const loop = this.createLoopRecord(id, options);

      await this.saveLoop(loop);
      this.startLoop(loop, 0);

      return ok(this.loopResult(loop));
    } catch (error) {
      return fail(error);
    }
  }

  async retomarLoops() {
    return this.resumeLoops();
  }

  async resumeLoops() {
    try {
      const loops = await this.readSavedLoops();

      for (const loop of loops) {
        this.startLoop(loop, 0);
      }

      return ok({
        retomados: loops.length,
        ids: loops.map((loop) => loop.pagamentoId)
      });
    } catch (error) {
      return fail(error);
    }
  }

  async cancelarAcompanhamento(id) {
    return this.cancelWatch(id);
  }

  async cancelWatch(id) {
    try {
      this.validateId(id);
      const key = String(id);
      const active = this.loops.get(key);

      if (active?.timer) clearTimeout(active.timer);

      this.loops.delete(key);
      await this.deleteLoopById(key);

      return ok({ id: key, cancelado: true });
    } catch (error) {
      return fail(error);
    }
  }

  async checkCharge(id) {
    try {
      this.validateId(id);
      const response = await this.http.get(`/v1/payments/${encodeURIComponent(id)}`);

      return ok(normalizePayment(response, this.config.sandbox));
    } catch (error) {
      return fail(error);
    }
  }

  interpretarWebhook(input) {
    return this.parseWebhook(input);
  }

  parseWebhook(input) {
    try {
      verifySignature({
        headers: input.headers,
        query: input.query,
        body: input.body || input,
        secret: this.config.webhookSecret
      });

      return ok(parseWebhook(input));
    } catch (error) {
      return fail(error);
    }
  }

  async processarWebhook(input) {
    return this.handleWebhook(input);
  }

  async handleWebhook(input) {
    const parsed = this.parseWebhook(input);

    if (!parsed.ok) {
      this.emit('erro', parsed);
      return parsed;
    }

    const payment = await this.checkCharge(parsed.dados.id);

    if (!payment.ok) {
      this.emit('erro', payment);
      return payment;
    }

    const eventName = paymentEventName(payment.dados.status);
    this.emit(eventName, payment.dados);

    if (eventName === 'desconhecido') {
      const error = fail(createError('STATUS_DESCONHECIDO', { details: payment.dados }));
      this.emit('erro', error);
    }

    return payment;
  }

  toPreferencePayload(charge) {
    const payload = {
      items: [
        {
          id: String(charge.productId),
          title: charge.productName,
          quantity: Number(charge.quantity),
          unit_price: Number(charge.price),
          currency_id: charge.currency
        }
      ],
      external_reference: String(charge.productId)
    };

    const notificationUrl = charge.webhookUrl || this.config.webhookUrl;
    if (notificationUrl) payload.notification_url = notificationUrl;
    if (charge.metadata) payload.metadata = charge.metadata;
    if (this.config.sandbox) payload.metadata = this.toTestMetadata(payload.metadata);
    if (charge.payerEmail) payload.payer = { email: charge.payerEmail };

    const backUrls = this.toBackUrls(charge);
    if (backUrls) payload.back_urls = backUrls;

    return payload;
  }

  toPixPayload(charge) {
    const payload = {
      transaction_amount: Number(charge.price) * Number(charge.quantity),
      description: charge.productName,
      payment_method_id: 'pix',
      external_reference: String(charge.productId),
      payer: {
        email: charge.payerEmail
      }
    };

    const notificationUrl = charge.webhookUrl || this.config.webhookUrl;
    if (notificationUrl) payload.notification_url = notificationUrl;
    if (charge.metadata) payload.metadata = charge.metadata;
    if (this.config.sandbox) payload.metadata = this.toTestMetadata(payload.metadata);

    return payload;
  }

  toRecurringPixPayload(subscription) {
    const payload = {
      reason: subscription.productName,
      external_reference: String(subscription.productId),
      payer_email: subscription.payerEmail,
      auto_recurring: {
        frequency: Number(subscription.frequency),
        frequency_type: subscription.frequencyType,
        transaction_amount: this.roundMoney(Number(subscription.price) * Number(subscription.quantity)),
        currency_id: subscription.currency
      },
      back_url: subscription.returnUrl,
      status: 'pending'
    };

    if (subscription.startDate) payload.auto_recurring.start_date = subscription.startDate;
    if (subscription.endDate) payload.auto_recurring.end_date = subscription.endDate;
    const notificationUrl = subscription.webhookUrl || this.config.webhookUrl;
    if (notificationUrl) payload.notification_url = notificationUrl;
    if (subscription.metadata) payload.metadata = subscription.metadata;
    if (this.config.sandbox) payload.metadata = this.toTestMetadata(payload.metadata);

    return payload;
  }

  toBackUrls(charge) {
    const urls = {};
    if (charge.successUrl) urls.success = charge.successUrl;
    if (charge.failureUrl) urls.failure = charge.failureUrl;
    if (charge.pendingUrl) urls.pending = charge.pendingUrl;

    return Object.keys(urls).length ? urls : null;
  }

  toTestMetadata(metadata) {
    if (!metadata || typeof metadata !== 'object' || Array.isArray(metadata)) {
      return { mp_facil_modo_teste: true };
    }

    return {
      ...metadata,
      mp_facil_modo_teste: true
    };
  }

  normalizeSplit(input, charge) {
    const split = input?.split || input?.divisao || {};
    const contaAInput = split.contaA || split.accountA || input?.contaA || input?.accountA || {};
    const contaBInput = split.contaB || split.accountB || input?.contaB || input?.accountB || {};
    const contaAPercentual = Number(contaAInput.porcentagem ?? contaAInput.percentual ?? contaAInput.percentage);
    const contaBPercentual = Number(contaBInput.porcentagem ?? contaBInput.percentual ?? contaBInput.percentage);
    const accessToken = contaAInput.accessToken || contaAInput.access_token || contaAInput.apiKey || contaAInput.api_key;
    const marketplace = split.marketplace || input?.marketplace || this.marketplaceId();

    if (!accessToken) throw createError('SPLIT_CONTA_A_INVALIDA');
    if (Number.isNaN(contaBPercentual)) throw createError('SPLIT_CONTA_B_INVALIDA');
    if (Number.isNaN(contaAPercentual) || contaAPercentual + contaBPercentual !== 100) {
      throw createError('SPLIT_PERCENTUAL_INVALIDO');
    }

    const total = Number(charge.price) * Number(charge.quantity);
    const marketplaceFee = this.roundMoney(total * (contaBPercentual / 100));
    const valorEstimadoA = this.roundMoney(total - marketplaceFee);

    if (marketplaceFee < 0 || marketplaceFee > total) {
      throw createError('SPLIT_VALOR_INVALIDO');
    }

    return {
      contaA: {
        accessToken,
        percentual: contaAPercentual,
        valorEstimado: valorEstimadoA
      },
      contaB: {
        percentual: contaBPercentual
      },
      marketplace,
      marketplaceFee
    };
  }

  marketplaceId() {
    if (!this.config.appId) return undefined;

    return `MP-MKT-${this.config.appId}`;
  }

  roundMoney(value) {
    return Math.round((value + Number.EPSILON) * 100) / 100;
  }

  idempotencyKey(input, charge) {
    return input?.idempotencyKey || input?.chaveIdempotencia || `mp-facil-${charge.productId}-${Date.now()}`;
  }

  createLoopRecord(id, options = {}) {
    const intervaloMs = Number(options.intervaloMs ?? options.intervalMs ?? options.intervalo ?? 10000);
    const duracaoValue = options.duracaoMs ?? options.durationMs ?? options.duracao ?? options.duration;
    const duracaoMs = duracaoValue === undefined || duracaoValue === null ? null : Number(duracaoValue);

    if (Number.isNaN(intervaloMs) || intervaloMs <= 0) {
      throw createError('LOOP_INTERVALO_INVALIDO');
    }

    if (duracaoMs !== null && (Number.isNaN(duracaoMs) || duracaoMs <= 0)) {
      throw createError('LOOP_DURACAO_INVALIDA');
    }

    const agora = Date.now();
    const pagamentoId = String(id);

    return {
      id: this.loopId(pagamentoId),
      pagamentoId,
      status: 'monitorando',
      intervaloMs,
      duracaoMs,
      criadoEm: new Date(agora).toISOString(),
      iniciadoEmMs: agora,
      terminaEmMs: duracaoMs ? agora + duracaoMs : null,
      tentativas: 0,
      ultimoResultado: null,
      modoTeste: this.config.sandbox
    };
  }

  loopResult(loop) {
    return {
      id: loop.pagamentoId,
      status: loop.status,
      intervaloMs: loop.intervaloMs,
      duracaoMs: loop.duracaoMs,
      tentativas: loop.tentativas,
      modoTeste: loop.modoTeste,
      arquivo: this.loopFile(loop.pagamentoId)
    };
  }

  startLoop(loop, delay = loop.intervaloMs) {
    const pagamentoId = String(loop.pagamentoId);
    const active = this.loops.get(pagamentoId);

    if (active?.timer) clearTimeout(active.timer);

    const timer = setTimeout(() => {
      this.runLoop(loop);
    }, delay);

    timer.unref?.();
    this.loops.set(pagamentoId, { loop, timer });
  }

  async runLoop(loop) {
    const pagamentoId = String(loop.pagamentoId);

    if (this.loopExpired(loop)) {
      const error = fail(createError('LOOP_TEMPO_ESGOTADO', { details: this.loopResult(loop) }));
      this.emit('erro', error);
      await this.finishLoop(loop);
      return;
    }

    loop.tentativas += 1;
    const payment = await this.checkCharge(pagamentoId);
    loop.ultimoResultado = payment;

    if (!payment.ok) {
      this.emit('erro', payment);
      await this.saveLoop(loop);
      this.startLoop(loop);
      return;
    }

    const eventName = paymentEventName(payment.dados.status);

    if (eventName === 'pago' || eventName === 'recusado') {
      loop.status = eventName;
      await this.finishLoop(loop);
      this.emit(eventName, payment.dados);
      return;
    }

    if (eventName === 'desconhecido') {
      const error = fail(createError('STATUS_DESCONHECIDO', { details: payment.dados }));
      loop.status = 'erro';
      await this.finishLoop(loop);
      this.emit('erro', error);
      return;
    }

    this.emit(eventName, payment.dados);
    await this.saveLoop(loop);
    this.startLoop(loop);
  }

  loopExpired(loop) {
    return Boolean(loop.terminaEmMs && Date.now() >= loop.terminaEmMs);
  }

  async finishLoop(loop) {
    const pagamentoId = String(loop.pagamentoId);
    const active = this.loops.get(pagamentoId);

    if (active?.timer) clearTimeout(active.timer);

    this.loops.delete(pagamentoId);
    await this.deleteLoopById(pagamentoId);
  }

  async saveLoop(loop) {
    await fs.mkdir(this.memoryLoopPath, { recursive: true });
    await fs.writeFile(this.loopFile(loop.pagamentoId), JSON.stringify(loop, null, 2));
  }

  async readSavedLoops() {
    try {
      const files = await fs.readdir(this.memoryLoopPath);
      const jsonFiles = files.filter((file) => file.endsWith('.json'));
      const loops = [];

      for (const file of jsonFiles) {
        const content = await fs.readFile(path.join(this.memoryLoopPath, file), 'utf8');
        const loop = JSON.parse(content);

        if (loop?.pagamentoId && loop.status === 'monitorando') {
          loops.push(loop);
        }
      }

      return loops;
    } catch (error) {
      if (error.code === 'ENOENT') return [];
      throw error;
    }
  }

  async deleteLoopById(id) {
    try {
      await fs.unlink(this.loopFile(id));
    } catch (error) {
      if (error.code !== 'ENOENT') throw error;
    }
  }

  loopFile(id) {
    return path.join(this.memoryLoopPath, `${this.loopId(id)}.json`);
  }

  loopId(id) {
    return String(id).replace(/[^a-z0-9_.-]/gi, '_');
  }

  validateConfig() {
    if (!this.config.apiKey) throw createError('API_KEY_AUSENTE');

    if (this.config.appId !== undefined && String(this.config.appId).trim() === '') {
      throw createError('APP_ID_INVALIDO');
    }
  }

  validateCharge(charge) {
    if (!charge) throw createError('PAYLOAD_INCOMPLETO');
    if (!charge.productName) throw createError('PRODUTO_AUSENTE');
    if (charge.price === undefined || charge.price === null) throw createError('PRECO_AUSENTE');
    if (Number(charge.price) <= 0 || Number.isNaN(Number(charge.price))) throw createError('PRECO_INVALIDO');
    if (!charge.productId) charge.productId = crypto.randomUUID();
  }

  validateId(id) {
    if (!id) throw createError('ID_AUSENTE');
  }

  validatePayerEmail(email) {
    if (!email) throw createError('EMAIL_PAGADOR_AUSENTE');
    if (!/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(String(email))) {
      throw createError('EMAIL_PAGADOR_INVALIDO');
    }
  }

  validateRecurringPix(subscription) {
    this.validateCharge(subscription);
    this.validatePayerEmail(subscription.payerEmail);

    if (!subscription.returnUrl) throw createError('URL_RETORNO_AUSENTE');

    const frequency = Number(subscription.frequency);

    if (Number.isNaN(frequency) || frequency <= 0) {
      throw createError('FREQUENCIA_ASSINATURA_INVALIDA');
    }
  }
}

const MercadoPago = MercadoPagoSimplificado;
const MPFacil = MercadoPagoSimplificado;

export { MercadoPago, MPFacil };
export default MercadoPagoSimplificado;
