// mp-facil/src/normalizers.js
function pick(object, keys) {
  for (const key of keys) {
    if (object && object[key] !== undefined && object[key] !== null) {
      return object[key];
    }
  }

  return undefined;
}

export function normalizeConfig(config = {}) {
  return {
    apiKey: pick(config, ['apiKey', 'api_key', 'accessToken', 'access_token']),
    appId: pick(config, ['appId', 'id_app', 'applicationId', 'application_id']),
    webhookUrl: pick(config, ['webhookUrl', 'webhook', 'urlWebhook', 'notificationUrl', 'notification_url']),
    webhookSecret: pick(config, ['webhookSecret', 'segredoWebhook', 'secret', 'webhook_secret']),
    sandbox: Boolean(pick(config, ['sandbox', 'modoTeste'])),
    baseUrl: pick(config, ['baseUrl', 'base_url']),
    fetchImpl: pick(config, ['fetch', 'fetchImpl']),
    memoryLoopPath: pick(config, ['memoryLoopPath', 'pastaMemoryLoop', 'memory_loop_path']),
    autoResumeLoops: pick(config, ['autoResumeLoops', 'autoRetomarLoops'])
  };
}

export function normalizeCharge(input = {}) {
  return {
    productName: pick(input, ['produto', 'product', 'productName', 'nome', 'title']),
    productId: pick(input, ['id', 'produtoId', 'productId', 'externalReference', 'external_reference']),
    price: pick(input, ['preco', 'price', 'valor', 'amount']),
    quantity: pick(input, ['quantidade', 'quantity']) || 1,
    currency: pick(input, ['moeda', 'currency', 'currencyId', 'currency_id']) || 'BRL',
    webhookUrl: pick(input, ['webhookUrl', 'webhook', 'urlWebhook', 'notificationUrl', 'notification_url']),
    successUrl: pick(input, ['successUrl', 'urlSucesso', 'success_url']),
    failureUrl: pick(input, ['failureUrl', 'urlFalha', 'failure_url']),
    pendingUrl: pick(input, ['pendingUrl', 'urlPendente', 'pending_url']),
    payerEmail: pick(input, ['payerEmail', 'emailPagador', 'payer_email']),
    metadata: pick(input, ['metadata', 'metadados']),
    raw: pick(input, ['raw', 'bruto'])
  };
}

export function normalizePayment(data = {}, modoTeste = false) {
  return {
    id: data.id,
    status: data.status,
    statusDetalhado: data.status_detail,
    situacao: translateStatus(data.status),
    valor: data.transaction_amount,
    moeda: data.currency_id,
    descricao: data.description,
    referencia: data.external_reference,
    metadados: data.metadata || {},
    metodoPagamento: data.payment_method_id,
    tipoPagamento: data.payment_type_id,
    aprovadoEm: data.date_approved,
    criadoEm: data.date_created,
    atualizadoEm: data.date_last_updated,
    modoTeste,
    bruto: data
  };
}

export function normalizePreference(data = {}, modoTeste = false) {
  return {
    id: data.id,
    referencia: data.external_reference,
    link: modoTeste ? data.sandbox_init_point : data.init_point,
    linkProducao: data.init_point,
    linkTeste: data.sandbox_init_point,
    criadoEm: data.date_created,
    modoTeste,
    bruto: data
  };
}

export function normalizePixPayment(data = {}, modoTeste = false) {
  const transactionData = data.point_of_interaction?.transaction_data || {};

  return {
    id: data.id,
    status: data.status,
    statusDetalhado: data.status_detail,
    situacao: translateStatus(data.status),
    valor: data.transaction_amount,
    moeda: data.currency_id,
    descricao: data.description,
    referencia: data.external_reference,
    metadados: data.metadata || {},
    metodoPagamento: data.payment_method_id,
    tipoPagamento: data.payment_type_id,
    qrCode: transactionData.qr_code,
    qrCodeBase64: transactionData.qr_code_base64,
    ticketUrl: transactionData.ticket_url,
    copiaECola: transactionData.qr_code,
    expiraEm: data.date_of_expiration,
    criadoEm: data.date_created,
    atualizadoEm: data.date_last_updated,
    modoTeste,
    bruto: data
  };
}

export function normalizeRecurringPix(input = {}) {
  return {
    productName: pick(input, ['produto', 'product', 'productName', 'nome', 'title']),
    productId: pick(input, ['id', 'produtoId', 'productId', 'externalReference', 'external_reference']),
    price: pick(input, ['preco', 'price', 'valor', 'amount']),
    quantity: pick(input, ['quantidade', 'quantity']) || 1,
    currency: pick(input, ['moeda', 'currency', 'currencyId', 'currency_id']) || 'BRL',
    payerEmail: pick(input, ['payerEmail', 'emailPagador', 'payer_email']),
    returnUrl: pick(input, ['urlRetorno', 'returnUrl', 'backUrl', 'back_url', 'successUrl', 'urlSucesso']),
    frequency: pick(input, ['frequencia', 'frequency']) || 1,
    frequencyType: pick(input, ['tipoFrequencia', 'frequencyType', 'frequency_type']) || 'months',
    startDate: pick(input, ['dataInicio', 'startDate', 'start_date']),
    endDate: pick(input, ['dataFim', 'endDate', 'end_date']),
    webhookUrl: pick(input, ['webhookUrl', 'webhook', 'urlWebhook', 'notificationUrl', 'notification_url']),
    metadata: pick(input, ['metadata', 'metadados'])
  };
}

export function normalizeRecurringPixSubscription(data = {}, modoTeste = false) {
  const recurring = data.auto_recurring || {};

  return {
    id: data.id,
    status: data.status,
    situacao: translateSubscriptionStatus(data.status),
    descricao: data.reason,
    referencia: data.external_reference,
    emailPagador: data.payer_email,
    link: data.init_point,
    valor: Number(recurring.transaction_amount),
    moeda: recurring.currency_id,
    frequencia: recurring.frequency,
    tipoFrequencia: recurring.frequency_type,
    iniciaEm: recurring.start_date,
    terminaEm: recurring.end_date,
    proximaCobrancaEm: data.next_payment_date,
    urlRetorno: data.back_url,
    webhookUrl: data.notification_url,
    criadoEm: data.date_created,
    atualizadoEm: data.last_modified,
    modoTeste,
    bruto: data
  };
}

export function translateStatus(status) {
  const statuses = {
    approved: 'pago',
    authorized: 'pendente',
    in_process: 'pendente',
    pending: 'pendente',
    rejected: 'recusado',
    cancelled: 'recusado',
    refunded: 'recusado',
    charged_back: 'recusado'
  };

  return statuses[status] || 'desconhecido';
}

function translateSubscriptionStatus(status) {
  const statuses = {
    pending: 'pendente',
    authorized: 'ativa',
    paused: 'pausada',
    cancelled: 'cancelada',
    canceled: 'cancelada'
  };

  return statuses[status] || 'desconhecido';
}
