// mp-facil/src/webhook.js
import crypto from 'node:crypto';
import { createError } from './errors.js';
import { translateStatus } from './normalizers.js';

export function parseWebhook(input = {}) {
  const body = input.body || input;
  const query = input.query || {};
  const headers = normalizeHeaders(input.headers || {});
  const dataId = getDataId(body, query);
  const type = body.type || body.topic || query.topic;
  const action = body.action || query.action;

  if (!dataId || type !== 'payment') {
    throw createError('WEBHOOK_INVALIDO');
  }

  return {
    id: dataId,
    tipo: type,
    acao: action,
    evento: action || type,
    headers,
    query,
    corpo: body
  };
}

export function verifySignature({ headers = {}, query = {}, body = {}, secret }) {
  if (!secret) return true;

  const normalizedHeaders = normalizeHeaders(headers);
  const signature = normalizedHeaders['x-signature'];
  const requestId = normalizedHeaders['x-request-id'];
  const dataId = getDataId(body, query);

  if (!signature || !requestId || !dataId) {
    throw createError('ASSINATURA_INVALIDA');
  }

  const parts = parseSignature(signature);

  if (!parts.ts || !parts.v1) {
    throw createError('ASSINATURA_INVALIDA');
  }

  const manifest = buildManifest({ dataId, requestId, timestamp: parts.ts });
  const expected = crypto.createHmac('sha256', secret).update(manifest).digest('hex');

  if (!safeEqual(expected, parts.v1)) {
    throw createError('ASSINATURA_INVALIDA');
  }

  return true;
}

export function paymentEventName(status) {
  return translateStatus(status);
}

function getDataId(body, query) {
  const value = query['data.id'] || query.id || body?.data?.id || body?.id;

  if (value === undefined || value === null) return undefined;

  return String(value).toLowerCase();
}

function parseSignature(signature) {
  return signature.split(',').reduce((parts, part) => {
    const [key, value] = part.split('=');

    if (key && value) parts[key.trim()] = value.trim();

    return parts;
  }, {});
}

export function buildManifest({ dataId, requestId, timestamp }) {
  return `id:${dataId};request-id:${requestId};ts:${timestamp};`;
}

function normalizeHeaders(headers) {
  return Object.entries(headers).reduce((normalized, [key, value]) => {
    normalized[key.toLowerCase()] = Array.isArray(value) ? value[0] : value;
    return normalized;
  }, {});
}

function safeEqual(expected, received) {
  const expectedBuffer = Buffer.from(expected);
  const receivedBuffer = Buffer.from(received);

  if (expectedBuffer.length !== receivedBuffer.length) return false;

  return crypto.timingSafeEqual(expectedBuffer, receivedBuffer);
}
